### PR TITLE
Update materials option + automatically install rhino3dm

### PR DESCRIPTION
**Pull request by tsvilans**
# Update materials option
Adds option to update previously imported materials.

## detailed explanation
* Adds update option (boolean) to 3dm import. Updates existing materials, otherwise creates new materials if existing ones are found. This allows making tweaks in Blender to materials and re-importing / updating the 3dm.

# Automatically install dependencies
Tries to import `rhino3dm` and installs it using `pip` if it fails.

## detailed explanation
* Tries to import `rhino3dm`, attempts to install it using `pip`
* Tries to use `pip`, installs it if it fails

### CODE REVIEW

**Comment by tsvilans** [comment]
Hi @jesterKing , tried to make a PR for the dependencies stuff and realized an older one was still open. Take a look and comment when you have a free moment...

**Comment by jesterKing** [comment]
Thanks, I'll check them when at my machine later today.

**Comment by jesterKing** [comment]
Oh! This one slipped my attention. As said, I'll check it later today. Thank you for the reminder, and sorry I forgot

File: import_3dm/read3dm.py [hunk 1]
```diff
@@ -22,10 +22,44 @@
...
-import rhino3dm as r3d
+
+def install_dependencies():
+    try:
+        try:
+            import pip
+        except:
+            from subprocess import call
+            print("Installing pip... "),
+            res = call("{} -m ensurepip".format(bpy.app.binary_path_python))
```
Comment (jesterKing) [inline import_3dm/read3dm.py]: Does this work for paths with spaces too? Or does binary_path_python already escape everything necessary?

Comment (tsvilans) [reply import_3dm/read3dm.py]: `binary_path_python` should handle escape everything, I think. It should be system specific, and handled by Blender.

**Comment by jesterKing** [comment]
With a fresh Blender build on Linux the dependency install isn't working yet. Investigating.

**Comment by jesterKing** [comment]
@tsvilans I have been trying to get this to work on my linux box as well. I got as far as having the pip dependency installed, but the pip install of rhino3dm fails.

The error that persists is: 

```Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: '/tmp/pip-req-tracker-m_hc2_0n/bcfbd6ee60f98835718df79377a8340417530d1465879ee428765018'```

This is the patch sofar. It improves the way pip dependency is satisfied so that it works at least also on my Linux box. - please review and apply the changes that also work still on your machine:

```diff --git a/import_3dm/read3dm.py b/import_3dm/read3dm.py
index b3cdfd9..ea92404 100644
--- a/import_3dm/read3dm.py
+++ b/import_3dm/read3dm.py
@@ -24,27 +24,44 @@ import os.path
 import bpy
 
 def install_dependencies():
+    import sys
+    import os
     try:
         try:
             import pip
         except:
             from subprocess import call
             print("Installing pip... "),
-            res = call("{} -m ensurepip".format(bpy.app.binary_path_python))
+            ensurepip = os.path.normpath(
+                    os.path.join(
+                        os.path.dirname(bpy.app.binary_path_python),
+                        "../lib/python{}.{}/ensurepip".format(
+                            sys.version_info.major,
+                            sys.version_info.minor)
+                        )
+                    )
+            res = call(bpy.app.binary_path_python, ensurepip)
 
             if res == 0:
                 import pip
             else:
                 raise Exception("Failed to install pip.")
 
-        print("Installing rhino3dm... "),
+        modulespath = os.path.normpath(
+                os.path.join(
+                    bpy.utils.script_path_user(),
+                    "addons/modules"
+                    )
+        if not os.path.exists(modulespath):
+           os.makedirs(modulespath) 
+        print("Installing rhino3dm to {}... ".format(modulespath)),
 
         try:
             from pip import main as pipmain
         except:
             from pip._internal import main as pipmain
 
-        res = pipmain(["install", "rhino3dm"])
+        res = pipmain(["install", "--upgrade", "--target", modulespath, "rhino3dm"])
         if res > 0:
             raise Exception("Failed to install rhino3dm.")
     except:
```

**Comment by jesterKing** [comment]
I just realise that in my code the os.sep should probably be used instead of forward slash.

**Comment by tsvilans** [comment]
Hey @jesterKing , did you try running Blender with administrator privileges? I noticed I needed those to install anything via `pip` through Blender.

**Comment by tsvilans** [comment]
> ```diff
> index b3cdfd9..ea92404 100644
> --- a/import_3dm/read3dm.py
> +++ b/import_3dm/read3dm.py
> @@ -24,27 +24,44 @@ import os.path
>  import bpy
>  
>  def install_dependencies():
> +    import sys
> +    import os
>      try:
>          try:
>              import pip
>          except:
>              from subprocess import call
>              print("Installing pip... "),
> -            res = call("{} -m ensurepip".format(bpy.app.binary_path_python))
> +            ensurepip = os.path.normpath(
> +                    os.path.join(
> +                        os.path.dirname(bpy.app.binary_path_python),
> +                        "../lib/python{}.{}/ensurepip".format(
> +                            sys.version_info.major,
> +                            sys.version_info.minor)
> +                        )
> +                    )
> ```


This expands to something liketh `2.80/python/lib/python3.7/ensurepip`. On Windows, there is no `python3.7` folder or similar.. the `libs` folder contains `ensurepip` directly. Should we put a OS check?

Also, as per your last comment, could just be:
>                     os.path.join(
>                         os.path.dirname(bpy.app.binary_path_python),
>                         "..",
>                         "lib",
>                         "python{}.{}/ensurepip".format(
>                             sys.version_info.major,
>                             sys.version_info.minor)
>                      )
>

> ```diff
>             res = call(bpy.app.binary_path_python, ensurepip)
>  
>              if res == 0:
>                  import pip
>              else:
>                  raise Exception("Failed to install pip.")
>  
> -        print("Installing rhino3dm... "),
> +        modulespath = os.path.normpath(
> +                os.path.join(
> +                    bpy.utils.script_path_user(),
> +                    "addons/modules"
> +                    )
> +        if not os.path.exists(modulespath):
> +           os.makedirs(modulespath) 
> +        print("Installing rhino3dm to {}... ".format(modulespath)),
> ```

This is so that `rhino3dm` gets installed to a separate modules directory? I wasn't aware of that, I thought it always just went into the `site-packages` folder in the Python dir. Nice!

**Comment by tsvilans** [comment]
> I just realise that in my code the os.sep should probably be used instead of forward slash.

`os.path.join` takes a list too, so you could just do:

>                     os.path.join(
>                         os.path.dirname(bpy.app.binary_path_python),
>                         "..",
>                         "lib",
>                         "python{}.{}/ensurepip".format(
>                             sys.version_info.major,
>                             sys.version_info.minor)
>                      )
>

**Comment by jesterKing** [comment]
@tsvilans it is not necessary to run Blender with elevated access on Linux. The changes I propose also try to install in the user add-ons folder, which is either under %APPDATA% or under .config/, which is all writable by the user (or should be anyway).

I haven't managed to find a fix for the problem with the pip install command from Blender. On the command-line it all worked correctly.

**Comment by tsvilans** [comment]
> @tsvilans it is not necessary to run Blender with elevated access on Linux. The changes I propose also try to install in the user add-ons folder, which is either under %APPDATA% or under .config/, which is all writable by the user (or should be anyway).
> 
> I haven't managed to find a fix for the problem with the pip install command from Blender. On the command-line it all worked correctly.

@jesterKing Yeah, installing to the user add-ons / modules folder sounds much better. Good to keep extra modules separate from the Blender Python stuff.

Regarding the `pip install` problem, is this still only on Linux? 

I found some links to similar problems:
* Something about getting rid of the `dist-info` folder before installing: https://github.com/googleapis/google-cloud-python/issues/6647#issuecomment-462816835
* Trying with elevated permissions: https://github.com/googleapis/google-cloud-python/issues/6647#issuecomment-485990360
* Moving the `METADATA` file into the right directory: https://stackoverflow.com/questions/54778630/could-not-install-packages-due-to-an-environmenterror-errno-2-no-such-file-or

**Comment by tsvilans** [comment]
> which is either under %APPDATA% or under .config/, which is all writable by the user (or should be anyway).

@jesterKing Also, do you know if this is stored in Blender somewhere? If there was a variable similar to `bpy.app.binary_path_python`, that would be safer.

**Comment by tsvilans** [comment]
> > which is either under %APPDATA% or under .config/, which is all writable by the user (or should be anyway).
> 
> @jesterKing Also, do you know if this is stored in Blender somewhere? If there was a variable similar to `bpy.app.binary_path_python` - something like `bpy.app.user_data_directory` - that would be safer.

@jesterKing found it: `bpy.utils.resource_path('USER')`

Sorry for spamming you this morning...

**Comment by jesterKing** [comment]
> @jesterKing found it: `bpy.utils.resource_path('USER')`

I used `bpy.utils.script_path_user()` in my diff. I suppose `resource_path('USER')`  works too, maybe they even translate to the same.

> I found some links to similar problems

Thanks, going through them.

**Comment by tsvilans** [comment]
> I used bpy.utils.script_path_user() in my diff. I suppose resource_path('USER')  works too, maybe they even translate to the same.

My bad, I didn't catch this. Yep, you're already using the right one. Sorry!

**Comment by jesterKing** [comment]
To get this forward and finally merged, could you please put the automatic dependency installation code behind a platform specific check? For now it should be ok to have that only on Windows, where there is probably the most benefit gained anyway.

**Comment by tsvilans** [comment]
@jesterKing Ok, that should hopefully do it..

**Comment by jesterKing** [comment]
@tsvilans I am trying to get it to run, but on my windows box I see this error:

> res = call(bpy.app.binary_path_python, ensurepip)
Traceback (most recent call last):
  File "<blender_console>", line 1, in <module>
  File "C:\Users\Nathan\PortableApps\blender-2.80-windows64\2.80\python\lib\subprocess.py", line 304, in call
    with Popen(*popenargs, **kwargs) as p:
  File "C:\Users\Nathan\PortableApps\blender-2.80-windows64\2.80\python\lib\subprocess.py", line 658, in __init__
    raise TypeError("bufsize must be an integer")
TypeError: bufsize must be an integer

FWIW the current patch for read3dm.py I am working with is
```

@@ -22,31 +22,55 @@
 
 import os.path
 import bpy
 
 def install_dependencies():
+    import sys
+    import os
     try:
         try:
             import pip
         except:
             from subprocess import call
             print("Installing pip... "),
-            res = call("{} -m ensurepip".format(bpy.app.binary_path_python))
+            pyver = ""
+            if sys.platform != "win32":
+                pyver = "python{}.{}".format(
+                    sys.version_info.major,
+                    sys.version_info.minor
+                )
+
+            ensurepip = os.path.normpath(
+                os.path.join(
+                    os.path.dirname(bpy.app.binary_path_python),
+                    "..", "lib", pyver, "ensurepip"
+                )
+            )
+            res = call(bpy.app.binary_path_python, ensurepip)
 
             if res == 0:
                 import pip
             else:
                 raise Exception("Failed to install pip.")
 
-        print("Installing rhino3dm... "),
+        modulespath = os.path.normpath(
+            os.path.join(
+                bpy.utils.script_path_user(),
+                "addons",
+                "modules"
+            )
+        )
+        if not os.path.exists(modulespath):
+           os.makedirs(modulespath) 
+        print("Installing rhino3dm to {}... ".format(modulespath)),
 
         try:
             from pip import main as pipmain
         except:
             from pip._internal import main as pipmain
 
-        res = pipmain(["install", "rhino3dm"])
+        res = pipmain(["install", "--upgrade", "--target", modulespath, "rhino3dm"])
         if res > 0:
             raise Exception("Failed to install rhino3dm.")
     except:
         raise Exception("Failed to install dependencies. Please make sure you have pip installed.")
```

Does this work for you?

**Comment by jesterKing** [comment]
There, I managed to fix this and have `pip`  and `rhino3dm` nicely install  automatically on Windows.

We can use this as basis for getting something similar to work on Linux and MacOS.

@tsvilans thanks for your patch! And apologies it took so long to get it done.

**Pull request merged**